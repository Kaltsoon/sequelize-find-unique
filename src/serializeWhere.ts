import { WhereOptions } from 'sequelize';
import jsonStringify from 'json-stable-stringify';

import { isPrimitive } from './utils';

const parseWhere = (value: unknown): WhereOptions => {
  if (typeof value !== 'object' || value === null) {
    throw new Error('Where parameter must be an object');
  }

  const entries = Object.entries(value);

  for (const [key, value] of entries) {
    if (!isPrimitive(value)) {
      throw new Error(
        `Where parameter properties can only contain serializable primitive values. Found value of type "${typeof value}" in property "${key}"`,
      );
    }
  }

  return value as WhereOptions;
};

const serializeWhere = (maybeWhere: unknown): string => {
  const where = parseWhere(maybeWhere);

  return jsonStringify(where);
};

export default serializeWhere;
