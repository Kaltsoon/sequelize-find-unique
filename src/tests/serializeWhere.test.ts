import serializeWhere from '../serializeWhere';

describe('serializeFindUniqueOptions', () => {
  test('serializes primitive values correctly', () => {
    expect(
      serializeWhere({
        firstName: 'Kalle',
        lastName: 'Ilves',
      }),
    ).toMatchSnapshot();
  });

  test('throws error for non-primitive values', () => {
    expect(() =>
      serializeWhere({
        firstName: 'Kalle',
        lastName: 'Ilves',
        friends: ['elina'],
      }),
    ).toThrowErrorMatchingSnapshot();
  });
});
