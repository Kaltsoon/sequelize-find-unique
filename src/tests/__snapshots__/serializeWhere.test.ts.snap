// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeFindUniqueOptions serializes primitive values correctly 1`] = `"{\\"firstName\\":\\"Kalle\\",\\"lastName\\":\\"Ilves\\"}"`;

exports[`serializeFindUniqueOptions throws error for non-primitive values 1`] = `"Where parameter properties can only contain serializable primitive values. Found value of type \\"object\\" in property \\"friends\\""`;
